PermissionsService/Services/PermissionsStore.cs

```diff
@@ -319,111 +312,146 @@ private async Task<string> FetchHttpSourceDocument(string sourceUri)
+                // Get scopes for multiple request URLs
+                foreach (var url in requestUrls)
+                {
+                    try
+                    {
+                        if (string.IsNullOrEmpty(url))
+                            throw new InvalidOperationException("The request URL cannot be null or empty.");
 
-                var scopes = permissionsData.ScopesListTable.Values.OfType<JToken>()
-                                                    .SelectMany(x => x).OfType<JProperty>()
-                                                    .SelectMany(x => x.Value).OfType<JProperty>()
-                                                    .Where(x => x.Name.Equals(scopeType))
-                                                    .SelectMany(x => x.Value)
-                                                    .Values<string>().Distinct().ToList();
-                scopes.Remove("N/A");
+                        var scopesForUrl = GetScopesForRequestUrl(requestUrl: url,
+                                                                scopeType: scopeType,
+                                                                method: method,
+                                                                leastPrivilegeOnly: leastPrivilegeOnly);
 
-                scopesInfo = GetScopesInformation(scopesInformationDictionary, scopes, scopeType, true);
+                        if (scopesForUrl == null || !scopesForUrl.Any())
+                            throw new InvalidOperationException($"Permissions information for {url} were not found.");
 
-                _telemetryClient?.TrackTrace("Return all permissions",
-                                             SeverityLevel.Information,
-                                             _permissionsTraceProperties);
-            }
-            else // fetch permissions for a given request url and method
-            {
-                if (string.IsNullOrEmpty(method))
-                {
-                    throw new ArgumentNullException(nameof(method), "The HTTP method value cannot be null or empty.");
+                        scopes.AddRange(scopesForUrl);
+                    }
+                    catch (Exception exception)
+                    {
+                        errors.Add(new PermissionError()
+                        {
+                            Url = url,
+                            Message = exception.Message
+                        });
+                    }
                 }
+            }
 
-                requestUrl = CleanRequestUrl(requestUrl);
-
-                // Check if requestUrl is contained in our Url Template table
-                TemplateMatch resultMatch = permissionsData.UriTemplateMatcher.Match(new Uri(requestUrl, UriKind.RelativeOrAbsolute));
+            // Get consent display name and description
+            var scopesInfo = GetAdditionalScopesInformation(scopesInformationDictionary, scopes.DistinctBy(x => new { x.ScopeName, x.ScopeType }));
```

Comment:
```suggestion
            var scopesInfo = GetAdditionalScopesInformation(scopesInformationDictionary, scopes.DistinctBy(x => $"{x.ScopeName}{x.ScopeType}", StringComparison.OrdinalIgnoreCase));
```

Otherwise you're doing reference comparisons on anonymous objects which will always be different

Reply:
right, this ends up calling [Object.Equals()](https://learn.microsoft.com/en-us/dotnet/api/system.object.equals?view=net-7.0) which can have a bunch of size effects and relies on reflection (bad perf) which recursively does reference comparisons on the object structure.
Whereas a string is deterministic and will provide better perf